ECIPs/ECIP-1042.md

```diff
@@ -0,0 +1,63 @@
+    ECIP: 1042
+    Title: GASPAY opcode
+    Author: Igor Artamonov <splix@etcdevteam.com>
+    Status: Draft
+    Type: Standard
+    Created: 2018-03-08
+    
+# ECIP-1042 - GASPAY opcode    
+    
+# Abstract
+Add an operation to consume gas and add amount paid for that gas to token balance.
+
+# Motivation
+A useful Smart Contract (i.e. valuable) should expect small reward for its execution. A micropayment could be very 
+small a sub-cent payment but sending an additional value to a contract will make it more complicated and cost of such 
+transfer or contract execution could exceed requested reward.
+
+It’s proposed here to introduce a new EVM Operation that will allow a developer to consume a portion of transaction fee 
+(amount paid for a gas) and transfer it to current Smart Contract. 
+
+Using a gas payment provided with transaction provides flexibility and independence from market price of Ether. Gas Price 
+is dynamic value and supposed to represent fair market price of contract execution, therefore making micropayment value 
+proportional to const of transaction and more stable in long term. 
+
+# Specification
+Provide a new operation `GASPAY a` where `a` is amount of gas to consume, and `0x0 <= a <= 0x0fffff`. The payment for 
+that gas goes to contract address. As a return it puts consumed value (Wei) into the stack.
+
+The cost of the operation itself is equal to 1/4 of requested amount of Gas plus 5000 (`a/4 + 5000`). The payment for 
```

Comment:
Is this extra "gas" to contract paid a) at the end of the contract execution (like how miner gas works), or b) paid immediately at the `GASPAY` opcode? If it is a), does it do anything with refunded gas as well?

Reply:
you mean should it be paid if transaction has failed? no, only if succeeded. but "payment" should be registered at the moment of opcode